### PR TITLE
V2.1 - SASL Quality of Protection

### PR DESCRIPTION
**Pull request by Kwoin**
With Kerberos authentication, you may have to set the property _javax.security.sasl.qop_ in order to succeed authentication, depending of the security policy

This pull request includes : 

- Added a new optional element **saslQop** into lsc-core-2.1.xsd
- Added new line to JndiService to use this new setting

I did not add generated xjc class, maven will do in each build.

(Btw, I also added IntelliJ IDEA files in .gitignore)

### CODE REVIEW

File: src/main/java/org/lsc/jndi/JndiServices.java [hunk 1]
```diff
@@ -101,15 +58,21 @@
...
 
+import javax.naming.*;
+import javax.naming.directory.*;
+import javax.naming.ldap.*;
+import javax.security.auth.callback.*;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
```
Comment (rouazana) [inline src/main/java/org/lsc/jndi/JndiServices.java]: please don't reorganize import like this, we prefer to not have .* imports

File: .gitignore [hunk 2]
```diff
@@ -8,8 +8,4 @@
 *.iml
 
 # bin
-/target/
-
-# xjc generation
-src/main/java/org/lsc/configuration/
-lsc.episode
\ No newline at end of file
+/target/
```
Comment (rouazana) [inline .gitignore]: cannot you fix the first commit instead?

**Comment by rouazana** [comment]
BTW thank you for your contribution!

File: src/main/java/org/lsc/jndi/JndiServices.java (hunk 1 above)
Comment (Kwoin) [inline src/main/java/org/lsc/jndi/JndiServices.java]: Ah sry for that, I'll fix the IDE settings

File: .gitignore (hunk 2 above)
Comment (Kwoin) [inline .gitignore]: You may prefer I squash all the commit in one ?
(Sorry, I'm not totally comfortable with pull requests yet)

Comment (rouazana) [inline .gitignore]: nope, having multiple commits is fine. That's just your first and your third commit that should be squashed together. You can do this easily with git rebase -i

**Comment by Kwoin** [comment]
I've made required changes (squash commits and declutter imports, it seems my IDE still rearrange them a bit, I hope it does not affect your coding style too much)

**Comment by coudot** [comment]
@rouazana if this is ok for you, I let you apply the PR and merge it into master

Regarding documentation, I can have a look at it. I suppose only this page needs an update: http://lsc-project.org/documentation/howto/kerberos ?

**Comment by rouazana** [comment]
Sorry I won't have time to do the merge myself (it involves some non trivial rebasing I guess). So please do it, or if @Kwoin can, please provide PR a PR for master too.

**Comment by coudot** [comment]
@soisik do you think you can help use to merge this PR on trunk?

**Comment by soisik** [comment]
@coudot if not mistaking,you want this PL to be merged in master but not 2.1 branch, or both ?

**Comment by coudot** [comment]
This PR can be merged in 2.1 directly, but it seems to need some adaptations to be merged in trunk. So I would like to have both ;)

**Comment by soisik** [comment]
Done on both branches.

**Comment by coudot** [comment]
Thanks a lot!

Do you think we could add a unit test on this new parameter?

**Comment by coudot** [comment]
Documentation updated here : http://lsc-project.org/documentation/howto/kerberos#lsc

@Kwoin could you tell me if this is ok?

**Comment by Kwoin** [comment]
@coudot 
the _**saslQop**_ element should appear after **_tlsActivated_** element, so the XSD wont complain about the sequence order.

**Comment by soisik** [comment]
@coudot in documentation, you should mention that default value is "auth".

As for the junit test, do you want a full GSSAPI / kerberos auth test ? Should be possible (spring-security-kerberos has junit tests on kerberos auth). I don't see anything else to test around this feature.

**Comment by coudot** [comment]
@Kwoin thanks, doc updated
@soisik maybe a simple test is to have a sample lsc.xml with this attribute set to check that the configuration is valid. Testing Kerberos is more complex, could be interesting but not needed at short term.

**Pull request merged**